DRILL-5716: Queue-driven memory allocation

paul-rogers opened the pull request:
Please see [DRILL-5716](https://issues.apache.org/jira/browse/DRILL-5716) for an overview.

This PR provides five commits that implement a method of allocating memory to queries based on ZK-based queues. The key concept is that the system has a fixed amount of memory. Suppose all queries need the same amount of memory. In that case, if we want to run n queries, each gets an amount that is the total memory divided by n.

This raises two questions:

1. Queries don't need the same amount of memory, so how do we handle this?
2. How do we know how many queries, n, are to run within Drill?

The solution is to leverage an existing feature: the Zookeeper (ZK) based queueing mechanism. This mechanism limits (throttles) the number of active queries. If we set the limit to n, then we know that Drill won't run more than n queries, and we can apply the above math.

But, not all queries are the same size. The ZK-based throttling mechanism addressed this by allowing *two* queues: a "large query" queue and a "small query" queue. The idea is that the cluster might run a single large ETL query, but at the same time, run five or 10 small interactive queries. The numbers are configurable via system options. The existing mechanism uses query cost to decide when a query is "small" vs. "large." The user sets a threshold in terms of cost. Queries split into the two queues accordingly.

### Resource Manager Layer

The first commit introduces a new resource management layer with a number of concepts:

* A pluggable, global resource manager configured at boot time. The code offers three versions: the null manager (no throttling), the ZK-based one discussed above, and a test-only one used in unit tests that uses a simple in-process queue.
* A Per-query resource allocator created by the resource manager to work with a single query through the query lifecycle.

Since the default manager does no queueing, queues themselves are internal to the resource manager. The model assumes:

* A query queue that manages the Drillbit's view into the distributed queue state.
* A queue "lease" that represents a grant of a single slot within a queue to run a query.

The design handles the current resource managers, and allows creating custom schedulers as needed for specific distributions or applications.

Prior to this PR, the Foreman worked with the ZK-based queues directly inline in the Foreman code. With the above framework in place, this PR refactors the Foreman to extract the ZK-based queue code into a resource manager implementation. Functionality is identical, just the location of the code moves to allow a cleaner design.

One interesting design issue is worth pointing out. The user can enable and disable ZK-queues at run time. However, the resource manager is global. To make this work, a “dynamic” resource manager wraps the ZK-based implementation. The dynamic version periodically checks system options to see if the ZK-based version is enabled. The dynamic version calls to either the ZK-based version or the default (non-throttled) version depending on what it finds in the system options.

When a switch occurs, queries already queued will stay queued. Only new queries use the new setting. This provides a graceful transition to/from throttled mode.

### Specifics of Memory Planning

The core of this change is a new memory planner. Here's how it works.

* The user enables queues using the existing `exec.queue.enable` session option.
* The user decides the number of "small" vs. "large" queries to run concurrently by setting `exec.queue.small` and `exec.queue.large` respectively. (The original numbers for these settings were far to high, this PR changes the defaults to much lower values.)
* Decide how to split memory across the two queues by specifying the `exec.queue.memory_ratio` value. The default is 10, which means that large queries get 10 units of memory while small queries get 1 unit.
* Decide on the planner cost threshold that separates "small" vs. "large" queries using `exec.queue.threshold`.
* Decide on the maximum queue wait time by setting `exec.queue.timeout_millis`. (Default is five minutes.)
* Not all operators participate in the memory allocation system (exchanges, hash join and Top-N, for example, still make unlimited use of memory.) To reserve memory for these operators, choose a value for the new config parameter, `exec.queue.memory_reserve_ratio` which sets aside a slice of total memory "unmanaged" operators. The default is 20% testing will tell us if that value should be larger or smaller.

The memory planner then does the following:

* Determine the number of “memory units” as the number of concurrent small queries + the number of concurrent large queries * the memory ratio.
* Determine the amount of memory available by taking the total system memory and multiplying it by 1 - the reserve ratio.
* Suppose this is a small query. Compute actual memory as available memory / number of memory units.
* Traverse the plan. Find the buffering operators grouped by node. (Buffering operators, in this release, are sort and hash agg.)
* Suppose we find that, on node X, we have two sorts and a hash agg. This is three operators on that node.
* Determine the number of minor fragments running on that node. Suppose there are 10.
* Divide the per-node memory across the operators on each node and each minor fragment. Here, we divide our total memory by 3 * 10 = 30 to get the per-operator memory amount.

The above is more complex than the existing memory allocation utilities. The extra behavior handles a special case seen in several production environments: the case in which a query runs only in the root fragment on a single node, so that there is only a single minor fragment. The existing code still divides memory by the possible number of fragments, even if the actual number is 1. The new code gracefully handles this case and will give all memory to the single fragment in this case.

### Planner and Foreman Changes

Memory planning now depends on the resource manager. The default manager assigns memory using the existing `MemoryAllocationUtilities`. The ZK-based version uses the math outlined above.

As a result, the third commit reworks the parallelizers to shift memory planning into a separate memory planning stage managed by the resource manager. As a result, we must shift creation of the JSON form of the query plan until after memory allocation is done.

We assume that some future resource manager may learn of available memory only after a queue lease is granted. So, we shift memory planning until after a queue slot is available; but before launching fragments.

### Web UI Enhancements

Early testing revealed that a major limitation of the existing ZK-based queues is that they are opaque: the user can't tell what is happening. To address this, two UI changes were made:

* If ZK-based queues are enabled, information about them now appears on the main Web UI page for the Drillbit.
* The main query profile page shows, for each query, the query cost (used to select the "small" vs. "large" queue) and the name of the selected queue: "small" or "large."

In addition, system/session options are now sorted by name to make it easier to find the queue-related options.

The above required changes to the query profile definition Protobuf (in commit 2) as well as changes in the various "models" and "templates" used by the Web UI. Note that the new information also automatically appears in the corresponding JSON REST API calls.

### Code Cleanup

Commit 5 has a large number of code cleanup items found while doing the above work. These do not affect functionality.

### Limitations and Future Improvements

The change in this PR is far from perfect; but it is the best we can do given the limitations of the existing ZK-based queues. In particular:

* There is no way to determine the total queue length: the number of queries waiting in the queue.
* There is no good way to cancel a query stuck in the queue.
* JDBC/ODBC clients won’t know that their queries are queued, or be able to cancel queued queries.
* The two-queue model is highly simplistic; real applications need a variety of queues and the means to categorize queries into queues.
* The ZK-queues allow no prioritization.
* The queueing mechanism does not limit CPU use, only memory use.
* The planner number used to split queries into queues is not very accurate or intuitive.
* No provision is made to understand that some queries need little memory, others need a huge amount. Instead, all queries are either “small” or “large.”
* Several operators still make unlimited use of memory. It will be hard to set aside a proper reserve for these operators, somewhat limiting the usefulness of this feature until those operators are fixed.

All that said, the present PR is a (small) step in the right direction and lays the groundwork for additional, future efforts.

ppadma commented on exec/java-exec/src/main/java/org/apache/drill/exec/work/foreman/rm/DynamicResourceManager.java:
```diff
@@ -0,0 +1,113 @@
+  }
+
+  @Override
+  public synchronized QueryResourceAllocator newQueryPlanner(QueryContext queryContext) {
```
this synchronization will be a problem with high concurrency. Can we reduce the number of synchronizations we are doing per query ? Seems like you are doing when newQueryPlanner is called and newExecRM is called. Is one enough for both ?

paul-rogers replied:
The refresh is done periodically, not every time that `refreshRM()` is called.

The problem is, we don't know if we are the thread that will cause the periodic refresh, or if the thread right after us will do it. If another thread does it, we might end up building a plan based on variables that change.

This an probably be rewritten in a functional style: return an object with current settings, and create a new object with new settings when the refresh occurs. Synchronization is still needed unless Java pointer assignment is thread-safe; I'd need to check.

In any event, this whole mechanism is more for larger queries. If we can run at very high concurrency, then each query uses very little resources and throttling becomes less necessary.

paul-rogers commented on exec/java-exec/src/main/java/org/apache/drill/exec/ExecConstants.java:
```diff
@@ -373,13 +373,26 @@
+  // Ratio of memory for small queries vs. large queries.
+  // Each small query gets 1 unit, each large query gets QUEUE_MEMORY_RATIO units.
+
+  DoubleValidator QUEUE_MEMORY_RATIO = new RangeDoubleValidator("exec.queue.memory_ratio", 0.001, 1000);
```
Sure, makes sense.

paul-rogers commented on exec/java-exec/src/main/java/org/apache/drill/exec/server/rest/DrillRestServer.java:
```diff
@@ -186,9 +187,10 @@ public void dispose(WebUserConnection instance) {
     @Override
     public WebUserConnection provide() {
-      final HttpSession session = request.getSession();
+//      final HttpSession session = request.getSession();
```
Overabundance of caution? But, since nothing broke when I commented out the line, I've gone ahead and deleted it.

paul-rogers commented on exec/java-exec/src/main/java/org/apache/drill/exec/work/foreman/rm/DefaultResourceManager.java:
```diff
@@ -0,0 +1,122 @@
+    public void visitAbstractPlan(PhysicalPlan plan) {
+      boolean replanMemory = ! plan.getProperties().hasResourcePlan;
+      if (! replanMemory || plan == null) {
+        return;
```
Good catch! Not sure why the plan would be null... Anyway the `!plan` is C++, so adjusted your suggestion to fit Java semantics.

paul-rogers commented on exec/java-exec/src/main/java/org/apache/drill/exec/work/foreman/rm/DefaultResourceManager.java:
```diff
@@ -0,0 +1,122 @@
+    public void setCost(double cost) {
+      // Nothing to do. The EXECUTION option in Foreman calls this,
+      // but does not do the work to plan sort memory. Is EXECUTION
+      // even used?
```
Exactly, that is the point. There is a code path in the Foreman that figures out the cost, then does nothing with it. There seem to be multiple ways that we work with plans, and I struggled to understand them all.

Anyway, since the comment is not useful to anyone but me, I removed it.

paul-rogers commented on exec/java-exec/src/main/java/org/apache/drill/exec/work/foreman/rm/DefaultResourceManager.java:
```diff
@@ -0,0 +1,122 @@
+
+  public DefaultResourceManager() {
+    memoryPerNode = DrillConfig.getMaxDirectMemory();
+    cpusPerNode = Runtime.getRuntime().availableProcessors();
```
This just grabs the total actual CPUs. Code elsewhere would do calcs based on the number. Same way that the code here grabs total memory, but other code does calcs on that (such as setting aside a reserve.)

But, CPU allocation is not yet implemented so this code is a bit of a dead end at the moment. However, did not want to remove it 'cause I'd have to figure it out again later when we add CPU allocation.

paul-rogers commented on exec/java-exec/src/main/java/org/apache/drill/exec/work/foreman/rm/DefaultResourceManager.java:
```diff
@@ -0,0 +1,122 @@
+  public int cpusPerNode() { return cpusPerNode; }
+
+  @Override
+  public QueryResourceAllocator newQueryPlanner(QueryContext queryContext) {
```
Done. Was a vestige of an old class name.

paul-rogers commented on exec/java-exec/src/main/java/org/apache/drill/exec/work/foreman/rm/DefaultResourceManager.java:
```diff
@@ -0,0 +1,122 @@
+  }
+
+  @Override
+  public QueryResourceManager newExecRM(final Foreman foreman) {
```
Done.

paul-rogers commented on exec/java-exec/src/main/java/org/apache/drill/exec/work/foreman/rm/DistributedQueryQueue.java:
```diff
@@ -0,0 +1,272 @@
+      this.queryId = queryId;
+      this.queueName = queueName;
+      this.lease = lease;
+      this.queryMemory = queryMemory;
```
Subtle point here. The rules for computing the memory are based on the queue, but they can change at any moment. The value here captures the value calculated at the moment that the lease is granted and is used to plan memory for the query.

paul-rogers commented on exec/java-exec/src/main/java/org/apache/drill/exec/work/foreman/rm/DistributedQueryQueue.java:
```diff
@@ -0,0 +1,272 @@
+  }
+
+  @Override
+  public long getDefaultMemoryPerNode(double cost) {
```
Changed.

paul-rogers commented on exec/java-exec/src/main/java/org/apache/drill/exec/work/foreman/rm/DistributedQueryQueue.java:
```diff
@@ -0,0 +1,272 @@
+  /**
+   * This limits the number of "small" and "large" queries that a Drill cluster will run
+   * simultaneously, if queuing is enabled. If the query is unable to run, this will block
+   * until it can. Beware that this is called under run(), and so will consume a Thread
```
Excellent points! This is why the ZK-based queue mechanism can't be our ultimate solution.

A true scheduler should limit the queue length. But, note, we need to limit the *global* length. If the limit is 5 then the sixth query, *regardless of Drillbit*, should fail. But, with ZK queues, the best we can do is keep a per-Drillbit list of enqueued queries and have a per-node limit. This means that we'd use an unfair algorithm: whether your query is queued depends on whether you ended up on a busy Drillbit or not.

The justification, I suppose, for current behavior is this. Without queueing, all queries run and use the same amount of memory in the Foreman *plus* a large amount of memory on each node. At least, here, only Foreman memory is used (plus a Foreman thread.)

Note that the ZK-based mechanism does, in fact, have a queue timeout, so there is at least that safety-valve.

ppadma replied:
Agreed. This is better than current behavior in that sense. At least, we are limiting the impact  to only foreman node.

paul-rogers commented on exec/java-exec/src/main/java/org/apache/drill/exec/work/foreman/rm/DistributedQueryQueue.java:
```diff
@@ -0,0 +1,272 @@
+      logger.debug("Query {} with cost {} placed into the {} queue.",
+                   QueryIdHelper.getQueryId(queryId), cost, queueName);
+
+      lease = distributedSemaphore.acquire(queueTimeout, TimeUnit.MILLISECONDS);
```
Don't turn on queueing for high concurrency. This option should never be used for versions of Drill designed for very high query rates.

In such a high-concurrency scenario, the app using Drill might consider the in-process queue based on Java concurrency objects. This would be particularly useful if the high-concurrency case ran the query only in the Foreman Drillbit. Of course, if the goal is very high concurrency, then queueing is not very useful as its whole point is to limit concurrency (to avoid overloading memory and CPU.)

In testing, the ZK queueing overhead turned out to be (surprisingly) quite negligible for a normal, fully-distributed query. Probably the cost of one additional network request is trivial compared to the number of network hops needed to get a distributed query running across Drillbits.

paul-rogers commented on exec/java-exec/src/main/java/org/apache/drill/exec/work/foreman/rm/DistributedQueryQueue.java:
```diff
@@ -0,0 +1,272 @@
+      } catch (final InterruptedException e) {
+        // if we end up here, the loop will try again
+      } catch (final Exception e) {
+        logger.warn("Failure while releasing lease.", e);
```
If we can't connect to ZK, then we'll drop out of the ZK-based Drill cluster and the Drillbit will terminate.

To be honest, I don't know why the code is written this way; I picked it up and moved it from the Foreman so it is original code prior to my change. I suppose the idea is that we're holding onto a lease. If we fail to release it, then available query slots will decrease. Do that enough times and no queries will run.

So, seems fair to retry until either 1) the Drillbit exits, or 2) the lease is released.

Probably should add a check of the Drillbit state: have we entered shutdown state?

Once we shutdown, I hope ZK will notice that the lease holder is gone (because the corresponding ZK node is ephemeral) and will force the lease to be released.

paul-rogers replied:
Good point. Added a mechanism to detect when the Drillbit is shutting down and abandon the loop. However, at present, Drill does not have an idea of a shutdown state. Fortunately, however, DRILL-4286 (graceful shutdown) will provide such a state. We can wire up the mechanisms at that time.

paul-rogers commented on exec/java-exec/src/main/java/org/apache/drill/exec/work/foreman/rm/QueryQueue.java:
```diff
@@ -0,0 +1,129 @@
+
+  /**
+   * Queue a query. The method returns only when the query is admitted for
+   * execution. As a result, the calling thread may block up to the configured
```
See note above about the limitations of the ZK-based distributed queue.

As noted, we can implement a limit when using the in-process, per-Foreman queue, but that only works for the embedded or single-server case.

paul-rogers commented on exec/java-exec/src/main/java/org/apache/drill/exec/work/foreman/rm/ResourceManager.java:
```diff
@@ -0,0 +1,62 @@
+
+  /**
+   * Returns the memory, in bytes, assigned to each node in a Drill
+   * cluster. Since Drill nodes are symmetrical, knowing he memory on
```
Fixed.

paul-rogers commented on exec/java-exec/src/main/java/org/apache/drill/exec/work/foreman/rm/ThrottledResourceManager.java:
```diff
@@ -0,0 +1,305 @@
+  /**
+   * Per-query resource manager. Handles resources and optional queue lease for
+   * a single query. As such, this is a non-shared resource: it is associated with
+   * a foreman: a single tread at plan time, and a single event (in some thread)
```
Fixed.

paul-rogers commented on exec/java-exec/src/main/resources/drill-module.conf:
```diff
@@ -344,6 +357,20 @@ drill.exec: {
+    // limit *actual* CPU usage; only the amount of CPU assumed to exist when
+    // planning and managing queries. Primarily for testing.
+    cpus_per_node: 0,
+  }
```
Well, it is not for testing only. Suppose that Drill runs under YARN. Then Drill should not use the full set of CPUs; it should use the number of vcores granted by YARN. In this case, the YARN launcher can use `-Ddrill.exec.rm.cpus_per_node=5` to specify the number of vcores.

The memory per node is a safety-valve. At present, several operators (at least three) decline to participate in the memory assignment scheme. We may find it necessary to artificially limit memory in order to force certain queries to work if they end up using far more memory than expected. (True, the memory ratio runtime option also solves this issue...)

These options do not appear in `drill-override-example.conf`, so users won't know about them.

paul-rogers commented:
Responses to the general comments:

> I think we need to limit the number of queries waiting for queues.

Please see the detailed notes. The ZK-based mechanism does not (AFAIK) provide the means to limit queue size. This is why this feature is preliminary, and may be why it has been disabled, though it has been available in the code for several years.

> For short queries, this feature might add more overhead/latency (going to zookeeper, extra synchronization mechanisms added etc.). This will be a problem with high concurrency. Did we quantify that ?

This is not a feature for situations of very high concurrency (= very small queries). It is a feature for situations in which a cluster runs enough distributed queries that the cluster becomes overloaded. For the case with 1000s of small, single-node queries, a app could use the framework provided to implement an in-process queue. In fact, one exists, but it is intended for unit tests.

> Not being to cancel once enqueued is also a problem. Can we do that by having cancel release the semaphore ?

Yes, cancelation is an issue. Since the ZK-based queues were an existing feature, I didn't dive too deeply into how they handle query cancellation. This would be a good area for QA to exercise. 

> Also, does semaphore count give an indication how many are waiting in the queue ?

There is no ready means to determine this, though I suspect that something could be put together that queries the internal znode and counts the ephemeral nodes that represent waiting processes. Doing so would be handy to display queue size in the Drill web UI. This would be a good enhancement once we prove that the basic mechanism works.

paul-rogers commented:
@jinfengni, can you take a quick look at the Foreman changes? Especially the bits that muck about with the physical plan: some of the work is moved from one place to another. Thanks!

arina-ielchiieva commented:
@paul-rogers could you please resolve the conflicts?

paul-rogers commented:
Looks like significant changes where made to the UI since this PR was first posted. Will need to rework my changes in that area. Please review the other bits while that rework is in progress.

paul-rogers commented:
Rebased onto the latest master.

The profile UI in master had changed since this PR was first posted. Reworked the query overview information to use the much better style added recently in master.

* The "Queued" field for duration appears only if queues are enabled.
* The overview line includes query cost and queue if queuing is enabled.

arina-ielchiieva commented on exec/java-exec/src/main/resources/drill-module.conf:
```diff
@@ -392,10 +419,12 @@ drill.exec.options:  {
     exec.queue.enable: false,
-    exec.queue.large: 10,
-    exec.queue.small: 100,
+    exec.queue.large: 2,
```
Do we need to add description of this properties in `drill-override-example.conf`?
Also is there a reason why we made limit smaller than before?

paul-rogers commented on protocol/src/main/java/org/apache/drill/exec/proto/UserBitShared.java:
```diff
@@ -12140,6 +12140,31 @@ public Builder clearDef() {
     com.google.protobuf.ByteString
         getOptionsJsonBytes();
+
+    // optional double total_cost = 7;
```
I would, but this is generated code...

arina-ielchiieva commented on zookeeper.out:
```diff
@@ -0,0 +1,31 @@
+2017-10-05 20:21:32,418 [myid:] - INFO  [main:QuorumPeerConfig@124] - Reading configuration from: /Users/paulrogers/bin/zookeeper-3.4.9/bin/../conf/zoo.cfg
```
I guess this file was added by mistake...

paul-rogers replied:
Yup. Removed.

arina-ielchiieva commented on exec/java-exec/src/main/resources/rest/index.ftl:
```diff
@@ -74,23 +79,66 @@
-                  <td>Client to Bit Encryption:</td>
-                  <td>${model.isUserEncryptionEnabled()?string("enabled", "disabled")}</td>
+                  <td>Client to Bit Encryption</td>
+                  <td class="list-value"	>${model.isUserEncryptionEnabled()?string("Enabled", "Disabled")}</td>
```
Redundant spaces after `class="list-value"`

paul-rogers replied:
Fixed.

arina-ielchiieva commented on exec/java-exec/src/main/java/org/apache/drill/exec/work/foreman/rm/AbstractResourceManager.java:
```diff
@@ -0,0 +1,68 @@
+    // lower than actual memory). Primarily for testing.
+
+    long memLimit = DrillConfig.getMaxDirectMemory();
+    long configMemoryPerNode = config.getBytes(ExecConstants.MAX_MEMORY_PER_NODE);
```
Nice, thanks for explanation.

arina-ielchiieva commented:
@paul-rogers please don't forget to resolve conflicts.

arina-ielchiieva commented on exec/java-exec/src/main/java/org/apache/drill/exec/work/foreman/rm/DistributedQueryQueue.java:
```diff
@@ -0,0 +1,342 @@
+    }
+
+    @Override
+    public boolean equals(Object other) {
```
Well, this is disputable. You may leave it as is but I still disagree.

paul-rogers replied:
Looks like Drill convention is that `equals()` is to be used only for collections (where `hashCode()` is also needed.) To avoid the misuse of `equals()` in this use case, I renamed the method to `isSameAs()` with the argument being of the same type.

I understand the issue, "equality" (or "sameness") is a deep philosophical discussion. (Remember the Heraclitus quote: "No person ever steps in the same river twice, for it's not the same river and he/she is not the same person.") Here, I just want to avoid keeping two copies of a zillion variables. It is just a convenience to have two instances of a class instead...

paul-rogers commented on exec/java-exec/src/main/java/org/apache/drill/exec/work/foreman/rm/DistributedQueryQueue.java:
```diff
@@ -0,0 +1,342 @@
+    }
+  }
+
+  public interface StatusAdapter {
```
Right. I'm more focused on helping developer understanding vs. helping the compiler. Does Drill have a standard? I do see nested classes and interfaces elsewhere in Drill. Of course, they do occur in the standard library as well. If Drill (or Apache) has a restriction, then I should refactor the code to follow that restriction. (And, we should file a ticket to fix other files that are out of compliance.)

arina-ielchiieva replied:
Well, there is no standard. It's up to the developer where to place interfaces / inner classes. I guess the only principle the developer should follow is where it is more logically correct.

arina-ielchiieva commented on exec/java-exec/src/main/java/org/apache/drill/exec/work/foreman/rm/DistributedQueryQueue.java:
```diff
@@ -165,12 +165,24 @@ public ConfigSet(SystemOptionManager optionManager) {
+     * changed between the snapshots
+     */
+
+    public boolean isSameAs(ConfigSet otherSet) {
```
This is definitely a compromise :) Thanks for making the changes.
